File: LeetCode/Medium/0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:

        start = 0
        max_len = 0
        char_index_map = {} 
        
        for end, char in enumerate(s):            
            if char in char_index_map:
                idx = char_index_map[char]
                
                if idx >= start:
                    start = idx + 1
            
            current_length = end - start + 1
            max_len = max(max_len, current_length)
            
            char_index_map[char] = end
            
        return max_len

File: LeetCode/Medium/0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        ios_base::sync_with_stdio(false);
        cin.tie(NULL);
        
        int l = 0, max_len = 0;
        unordered_set<int> ust;
        for (int r = 0; r < s.size(); r++) {
            while (ust.contains(s[r])) {
                ust.erase(s[l]);
                l++;
            }
            ust.insert(s[r]);
            max_len = max(max_len, r - l + 1);
        }
        return max_len;
    }
};